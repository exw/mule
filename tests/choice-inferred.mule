# This is the classic example of something that hindley-milner can't give a
# type to, since choice is used at two different types in the body of the
# function. Mule should throw a type error here as well, but see
# choice-annotated.mule for a version that type checks.
fn choice.
  { x = choice {} {}
  , y = choice (fn x. x) (fn x. x)
  }
