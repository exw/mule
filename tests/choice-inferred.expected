
Desugared:

	fn $0. match-lam
| choice -> _ where { x = _ } (_ where { y = _ } {} (choice fn $2. match-lam
  | x -> x
  end $2 fn $1. match-lam
  | x -> x
  end $1)) (choice {} {})
end $0
Type error: mismatched type constructors: record and function
